src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -461,7 +463,15 @@ bool IsDerivedType(TypeSymbol possibleDerived, TypeSymbol possibleBase)
                 var tests = ArrayBuilder<Tests>.GetInstance(2);
                 var convertedInput = MakeConvertToType(input, constant.Syntax, constant.Value.Type!, isExplicitTest: false, tests);
                 output = convertedInput;
-                tests.Add(new Tests.One(new BoundDagValueTest(constant.Syntax, constant.ConstantValue, convertedInput)));
+                if (ValueSetFactory.ForInput(input)?.Related(BinaryOperatorKind.Equal, constant.ConstantValue).IsEmpty == true)
+                {
+                    // This could only happen for a length input where the permitted value domain (>=0) is a strict subset of possible values for the type (int)
+                    tests.Add(Tests.False.Instance);
+                }
```

Comment:
:memo: This wasn't checked for constant tests because it could never happen before.
We could possibly get by without this but we'd need to specifically cover it while rewriting nested length tests.